common/remote/sockfile.cpp

```diff
@@ -1989,6 +2011,104 @@ extern IFileIO *createRemoteFilteredFile(SocketEndpoint &ep, const char * filena
     return nullptr;
 }
 
+class CRemoteFilteredKeyIO : public CRemoteFilteredFileIOBase
+{
+public:
+    // Really a stream, but life (maybe) easier elsewhere if looks like a file
+    // Sometime should refactor to be based on ISerialStream instead - or maybe IRowStream.
+    CRemoteFilteredKeyIO(SocketEndpoint &ep, const char *filename, IOutputMetaData *actual, IOutputMetaData *projected, const RowFilter &fieldFilters, unsigned __int64 chooseN)
+        : CRemoteFilteredFileIOBase(ep, filename, actual, projected, fieldFilters, chooseN)
+    {
+        request.appendf(",\n \"kind\" : \"indexread\"");
+        closeRequest(request);
+        sendRequest(0, nullptr);
+    }
+};
+
+class CRemoteFilteredKeyCountIO : public CRemoteFilteredFileIOBase
+{
+public:
+    // Really a stream, but life (maybe) easier elsewhere if looks like a file
+    // Sometime should refactor to be based on ISerialStream instead - or maybe IRowStream.
+    CRemoteFilteredKeyCountIO(SocketEndpoint &ep, const char *filename, IOutputMetaData *actual, IOutputMetaData *projected, const RowFilter &fieldFilters, unsigned __int64 rowLimit)
+        : CRemoteFilteredFileIOBase(ep, filename, actual, projected, fieldFilters, rowLimit)
+    {
+        request.appendf(",\n \"kind\" : \"indexcount\"");
+        closeRequest(request);
+        sendRequest(0, nullptr);
+    }
+};
+
+class CRemoteKey : public CSimpleInterfaceOf<IIndexLookup>
+{
+    Owned<IFileIO> fileIO;
+    offset_t pos = 0;
+    Owned<ISourceRowPrefetcher> prefetcher;
+    CThorContiguousRowBuffer prefetchBuffer;
+    Owned<ISerialStream> strm;
+    bool pending = false;
+    SocketEndpoint ep;
+    StringAttr filename;
+    Linked<IOutputMetaData> actual, projected;
+    RowFilter fieldFilters;
+
+public:
+    CRemoteKey(SocketEndpoint &_ep, const char *_filename, IOutputMetaData *_actual, IOutputMetaData *_projected, const RowFilter &_fieldFilters, unsigned __int64 rowLimit)
+        : ep(_ep), filename(_filename), actual(_actual), projected(_projected)
+    {
+        for (unsigned f=0; f<_fieldFilters.numFilterFields(); f++)
+            fieldFilters.addFilter(OLINK(_fieldFilters.queryFilter(f)));
+        fileIO.setown(new CRemoteFilteredKeyIO(ep, filename, actual, projected, fieldFilters, rowLimit));
+        if (!fileIO)
+            throw MakeStringException(0, "Unable to open remote key part: '%s'", filename.get());
+        strm.setown(createFileSerialStream(fileIO));
+        prefetcher.setown(projected->createDiskPrefetcher());
+        if (prefetcher)
+            prefetchBuffer.setStream(strm);
+    }
+// IIndexLookup
+    virtual unsigned __int64 getCount() override
+    {
+        return checkCount(0);
+    }
+    virtual unsigned __int64 checkCount(unsigned __int64 limit) override
+    {
+        Owned<IFileIO> iFileIO = new CRemoteFilteredKeyCountIO(ep, filename, actual, projected, fieldFilters, limit);
```

Comment:
Is it correct to pass projected here?  What does it represent if this is for a standard index read?

Reply:
It won't actually be used. But will needlessly be serialized (if different from actual) to server.
I'll change.